draft-wang-ppm-differential-privacy.md

```diff
@@ -414,23 +459,204 @@ random seed into a sample from a given distribution.
+
+Client-DP allows Clients to protect their privacy by adding noise to their
+measurements directly, as described in {{levels}}. Analyses ({{FMT20}} and
+{{FMT22}}) have shown that the central DP guarantee can also be amplified
+by aggregating Clients' measurements with Client-DP.
+
+For this particular instantiation of `Prio3SumVec` with `bits = 1`, i.e. each
+value in the vector is either a 0 or 1, we can apply randomized response {{rr}}
+Client-DP to Clients' measurements, with a configured Client-DP with `EPSILON_0`
+parameter.
+
+The DP guarantee is met, as long as there are at least "minimum batch size"
+number of Clients, each of which adds the randomized response Client-DP.
+However, in case the minimum batch size is not met when Aggregators want to
+output their aggregate shares, each Aggregator can sample the same randomized
+response Client-DP on the "missing" Clients with zero-hot vectors, so the
+central DP guarantee can still be satisfied, based on the privacy amplification
+analysis in {{FMT20}} and {{FMT22}}.
+
+~~~
+class Prio3SumVecWithRandomizedResponse:
+    Measurement = List[int]
+    AggregateShare = List[Field]
+    AggregateResult = List[int]
+    DebiasedAggregateResult = List[float]
+
+    def __init__(self,
+                 rr_primitive: DpPrimitive,
+                 num_aggregators: Unsigned):
+        self.rr_primitive = rr_primitive
+        self.num_aggregators = num_aggregators
+
+    def maybe_add_noise_to_measurement(self,
+                                       meas: Measurement,
+                                       ) -> Measurement:
+        """
+        Apply randomized response primitive to measurement.
+        """
+        return self.rr_primitive.add_noise(meas)
+
+    def maybe_add_noise_to_agg_share(self,
+                                     agg_share: AggregateShare,
+                                     meas_count: Unsigned,
+                                     min_batch_size: Unsigned,
+                                     ) -> AggregateShare:
+        """
+        If minimum batch size is met, this function does nothing,
+        because DP guarantee would have already been met.
+        Otherwise, add the same randomized response noise on the
+        missing Clients.
+        """
+        if meas_count >= batch_size:
+            return agg_share
+        noise = self.rr_primitive.sample_noise(
+            batch_size - meas_count, len(agg_share)
+        )
+        return [a + Field(b) for (a, b) in zip(agg_share, noise)]
+
+    def maybe_debias(self,
+                     agg_result: AggregateResult,
+                     meas_count: Unsigned,
+                     min_batch_size: Unsigned,
+                     ) -> DebiasedAggregateResult:
+        """
+        Debias aggregate result with randomized response.
+        """
+        if meas_count >= min_batch_size:
+            return self.rr_primitive.debias(agg_result, meas_count)
+        # The count passed to debiasing will be:
```

Comment:
This is the tricky part about debiasing, when there is composition of Client-DP and Aggregator-DP.

Reply:
Let's use the terms "client noise" and "aggregator noise"; we don't want to conflate the "process" (adding noise) with the "result" (DP)

Reply:
I didn't explicitly explain Client-DP and Aggregator-DP noise in this function's comments. Is there anything else that I missed?

Reply:
LGTM